File: test/gc_alloc.c
/* Behavioral contract for the collector's memory: block allocation and the two
 * heap lists, the root set, the GC suspend counter, and the scratch allocator.
 * Run against whichever implementation the build selected (`-Dgc-alloc=c` or
 * the Zig default).
 *
 * Like the VM state contract, this one includes the internal headers. It has
 * to: every operation under test is a mutation of `janet_vm`'s garbage
 * collection fields, and the fields are the observable result. There is no
 * public accessor for `block_count` or `scratch_len`, and inventing one would
 * test the accessor.
 *
 * Two things are deliberately *not* exercised. Nothing here lets a synthetic
 * block reach `janet_sweep`: `test_gcalloc_*` unlinks what it allocated and
 * restores the counters, so the contract stays independent of the two
 * increments that still own marking and sweeping. And the two fatal paths --
 * `janet_srealloc` and `janet_sfree` on a pointer this allocator never handed
 * out -- abort the process, so they are described here rather than run.
 */

#include <assert.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "features.h"
#include <janet.h>
#include "state.h"
#include "gc.h"

/* ------------------------------------------------------------------ helpers */

static JanetScratch *header_of(void *mem) {
    return ((JanetScratch *) mem) - 1;
}

static int scratch_index_of(void *mem) {
    JanetScratch *s = header_of(mem);
    for (size_t i = 0; i < janet_vm.scratch_len; i++) {
        if (janet_vm.scratch_mem[i] == s) return (int) i;
    }
    return -1;
}

/* --------------------------------------------------------------- gcpressure */

/* The only thing janet_gcpressure does is move the threshold. It must not
 * collect, and it must not touch the block count -- the bytes it is told about
 * were allocated outside the collector's accounting. */
static void test_gcpressure(void) {
    size_t before = janet_vm.next_collection;
    size_t blocks = janet_vm.block_count;

    janet_gcpressure(0);
    assert(janet_vm.next_collection == before);

    janet_gcpressure(4096);
    assert(janet_vm.next_collection == before + 4096);
    assert(janet_vm.block_count == blocks);

    janet_vm.next_collection = before;
}

/* ------------------------------------------------------------------ gcalloc */

/* Undo one allocation, restoring every field it moved. Only valid for the
 * block at the head of its list, which is where janet_gcalloc just put it. */
static void unlink_head(int weak, size_t size) {
    JanetGCObject *head = weak ? janet_vm.weak_blocks : janet_vm.blocks;
    if (weak) {
        janet_vm.weak_blocks = head->data.next;
    } else {
        janet_vm.blocks = head->data.next;
    }
    janet_vm.block_count--;
    janet_vm.next_collection -= size;
    janet_free(head);
}

/* A new block goes on the front of the normal heap, carries its type in the
 * low byte of `flags` and nothing else, and is counted. It is emphatically not
 * marked: the caller has not filled it in yet, and a collection that treated it
 * as reachable would trace uninitialised memory. */
static void test_gcalloc_normal_heap(void) {
    size_t size = 128;
    void *previous = janet_vm.blocks;
    size_t count = janet_vm.block_count;
    size_t next = janet_vm.next_collection;
    void *weak = janet_vm.weak_blocks;

    JanetGCObject *mem = janet_gcalloc(JANET_MEMORY_ARRAY, size);
    assert(mem != NULL);
    assert(janet_vm.blocks == mem);
    assert(mem->data.next == previous);
    assert(janet_gc_type(mem) == JANET_MEMORY_ARRAY);
    assert(mem->flags == JANET_MEMORY_ARRAY);
    assert(!janet_gc_reachable(mem));
    assert(janet_vm.block_count == count + 1);
    assert(janet_vm.next_collection == next + size);
    assert(janet_vm.weak_blocks == weak);

    unlink_head(0, size);
    assert(janet_vm.blocks == previous);
    assert(janet_vm.block_count == count);
    assert(janet_vm.next_collection == next);
}

/* The four weak types are the ones at or above JANET_MEMORY_TABLE_WEAKK, and
 * the boundary is exactly that: the split is a numeric comparison against the
 * first weak constant, not a table of types. */
static void test_gcalloc_weak_heap(void) {
    enum JanetMemoryType weak_types[] = {
        JANET_MEMORY_TABLE_WEAKK,
        JANET_MEMORY_TABLE_WEAKV,
        JANET_MEMORY_TABLE_WEAKKV,
        JANET_MEMORY_ARRAY_WEAK
    };

    for (size_t i = 0; i < sizeof(weak_types) / sizeof(weak_types[0]); i++) {
        size_t size = 96;
        void *strong = janet_vm.blocks;
        void *previous = janet_vm.weak_blocks;
        size_t count = janet_vm.block_count;

        JanetGCObject *mem = janet_gcalloc(weak_types[i], size);
        assert(janet_vm.weak_blocks == mem);
        assert(mem->data.next == previous);
        assert(janet_gc_type(mem) == (int) weak_types[i]);
        assert(janet_vm.blocks == strong);
        assert(janet_vm.block_count == count + 1);

        unlink_head(1, size);
        assert(janet_vm.weak_blocks == previous);
        assert(janet_vm.block_count == count);
    }
}

/* Every type below the boundary goes on the normal heap. Worth stating for
 * JANET_MEMORY_NONE in particular, which is zero and therefore the value a
 * caller reaches by mistake. */
static void test_gcalloc_strong_types(void) {
    enum JanetMemoryType strong_types[] = {
        JANET_MEMORY_NONE,
        JANET_MEMORY_STRING,
        JANET_MEMORY_TABLE,
        JANET_MEMORY_FUNCDEF,
        JANET_MEMORY_THREADED_ABSTRACT
    };

    for (size_t i = 0; i < sizeof(strong_types) / sizeof(strong_types[0]); i++) {
        void *weak = janet_vm.weak_blocks;
        JanetGCObject *mem = janet_gcalloc(strong_types[i], 64);
        assert(janet_vm.blocks == mem);
        assert(janet_vm.weak_blocks == weak);
        assert(janet_gc_type(mem) == (int) strong_types[i]);
        unlink_head(0, 64);
    }
}

/* Successive allocations chain: the list is singly linked through the header's
 * `next`, newest first. */
static void test_gcalloc_chains(void) {
    void *previous = janet_vm.blocks;
    JanetGCObject *first = janet_gcalloc(JANET_MEMORY_NONE, 32);
    JanetGCObject *second = janet_gcalloc(JANET_MEMORY_NONE, 32);
    JanetGCObject *third = janet_gcalloc(JANET_MEMORY_NONE, 32);

    assert(janet_vm.blocks == third);
    assert(third->data.next == second);
    assert(second->data.next == first);
    assert(first->data.next == previous);

    unlink_head(0, 32);
    unlink_head(0, 32);
    unlink_head(0, 32);
    assert(janet_vm.blocks == previous);
}

/* --------------------------------------------------------------- root set */

/* Rooting appends. The root set is a multiset: n roots need n unroots. */
static void test_root_counting(void) {
    size_t base = janet_vm.root_count;
    JanetArray *a = janet_array(0);
    Janet v = janet_wrap_array(a);

    janet_gcroot(v);
    assert(janet_vm.root_count == base + 1);
    assert(janet_unwrap_array(janet_vm.roots[base]) == a);

    janet_gcroot(v);
    assert(janet_vm.root_count == base + 2);
    assert(janet_unwrap_array(janet_vm.roots[base + 1]) == a);

    assert(janet_gcunroot(v) == 1);
    assert(janet_vm.root_count == base + 1);
    assert(janet_gcunroot(v) == 1);
    assert(janet_vm.root_count == base);
    assert(janet_gcunroot(v) == 0);
    assert(janet_vm.root_count == base);
}

/* Roots are matched by pointer identity, not by value equality. Two arrays
 * with the same contents are different roots. */
static void test_root_identity_by_pointer(void) {
    size_t base = janet_vm.root_count;
    Janet a = janet_wrap_array(janet_array(0));
    Janet b = janet_wrap_array(janet_array(0));

    janet_gcroot(a);
    assert(janet_gcunroot(b) == 0);
    assert(janet_vm.root_count == base + 1);
    assert(janet_gcunroot(a) == 1);
    assert(janet_vm.root_count == base);
}

/* The three types the collector never traces compare equal to any value of
 * their own type. Rooting one number and unrooting a different one succeeds,
 * which is harmless -- the slot held nothing worth keeping either way -- but it
 * is observable, so it is pinned here. */
static void test_root_identity_of_immediates(void) {
    size_t base = janet_vm.root_count;

    janet_gcroot(janet_wrap_number(1.0));
    assert(janet_gcunroot(janet_wrap_number(9999.0)) == 1);
    assert(janet_vm.root_count == base);

    janet_gcroot(janet_wrap_true());
    assert(janet_gcunroot(janet_wrap_false()) == 1);
    assert(janet_vm.root_count == base);

    janet_gcroot(janet_wrap_nil());
    assert(janet_gcunroot(janet_wrap_nil()) == 1);
    assert(janet_vm.root_count == base);

    /* Different types never match, immediate or not. */
    janet_gcroot(janet_wrap_number(1.0));
    assert(janet_gcunroot(janet_wrap_true()) == 0);
    assert(janet_gcunroot(janet_wrap_nil()) == 0);
    assert(janet_gcunroot(janet_wrap_number(0.0)) == 1);
    assert(janet_vm.root_count == base);
}

/* Unrooting fills the vacated slot from the top of the set, so the order of
 * the remaining roots is not the order they were added in. */
static void test_root_removal_swaps_from_top(void) {
    size_t base = janet_vm.root_count;
    JanetArray *a = janet_array(0);
    JanetArray *b = janet_array(0);
    JanetArray *cc = janet_array(0);

    janet_gcroot(janet_wrap_array(a));
    janet_gcroot(janet_wrap_array(b));
    janet_gcroot(janet_wrap_array(cc));

    assert(janet_gcunroot(janet_wrap_array(a)) == 1);
    assert(janet_vm.root_count == base + 2);
    assert(janet_unwrap_array(janet_vm.roots[base]) == cc);
    assert(janet_unwrap_array(janet_vm.roots[base + 1]) == b);

    assert(janet_gcunroot(janet_wrap_array(b)) == 1);
    assert(janet_gcunroot(janet_wrap_array(cc)) == 1);
    assert(janet_vm.root_count == base);
}

/* janet_gcunrootall does not remove every rooting, despite what its comment in
 * gc.c says. It fills the vacated slot from the top and then advances, so the
 * value it just moved down is never examined: n rootings become floor(n / 2).
 * FOUND.md carries the defect; this pins the behaviour both implementations
 * have to produce. */
static void test_root_unrootall_halves(void) {
    size_t counts[] = {1, 2, 3, 4, 5, 8};

    for (size_t k = 0; k < sizeof(counts) / sizeof(counts[0]); k++) {
        size_t n = counts[k];
        size_t base = janet_vm.root_count;
        Janet v = janet_wrap_array(janet_array(0));

        for (size_t i = 0; i < n; i++) janet_gcroot(v);
        assert(janet_vm.root_count == base + n);

        assert(janet_gcunrootall(v) == 1);
        assert(janet_vm.root_count == base + n / 2);

        /* What survives really is still rooted, and can be removed one at a
         * time. */
        for (size_t i = 0; i < n / 2; i++) assert(janet_gcunroot(v) == 1);
        assert(janet_vm.root_count == base);
        assert(janet_gcunrootall(v) == 0);
    }
}

/* An absent value reports absence and changes nothing. */
static void test_root_unrootall_absent(void) {
    size_t base = janet_vm.root_count;
    Janet a = janet_wrap_array(janet_array(0));
    Janet b = janet_wrap_array(janet_array(0));

    janet_gcroot(a);
    assert(janet_gcunrootall(b) == 0);
    assert(janet_vm.root_count == base + 1);
    assert(janet_gcunroot(a) == 1);
    assert(janet_vm.root_count == base);
}

/* Growth is by doubling the required count, and the roots survive it. */
static void test_root_capacity_growth(void) {
    size_t base = janet_vm.root_count;
    Janet v = janet_wrap_array(janet_array(0));
    size_t added = 0;

    while (janet_vm.root_count < janet_vm.root_capacity) {
        janet_gcroot(v);
        added++;
    }
    size_t at_capacity = janet_vm.root_capacity;

    janet_gcroot(v);
    added++;
    assert(janet_vm.root_capacity == 2 * (at_capacity + 1));
    assert(janet_vm.root_count == base + added);
    for (size_t i = 0; i < added; i++) {
        assert(janet_unwrap_array(janet_vm.roots[base + i]) == janet_unwrap_array(v));
    }

    for (size_t i = 0; i < added; i++) assert(janet_gcunroot(v) == 1);
    assert(janet_vm.root_count == base);
}

/* ------------------------------------------------------------- suspension */

/* The handle is the depth to restore, not a token to match. Unlocking with an
 * outer handle discards every lock taken since, which is what makes it safe for
 * a cleanup path to hold one handle across nested regions. */
static void test_gclock_nesting(void) {
    int base = janet_vm.gc_suspend;

    int h1 = janet_gclock();
    assert(h1 == base);
    assert(janet_vm.gc_suspend == base + 1);

    int h2 = janet_gclock();
    assert(h2 == base + 1);
    assert(janet_vm.gc_suspend == base + 2);

    janet_gcunlock(h2);
    assert(janet_vm.gc_suspend == base + 1);
    janet_gcunlock(h1);
    assert(janet_vm.gc_suspend == base);

    h1 = janet_gclock();
    (void) janet_gclock();
    (void) janet_gclock();
    assert(janet_vm.gc_suspend == base + 3);
    janet_gcunlock(h1);
    assert(janet_vm.gc_suspend == base);
}

/* A suspended collector does not collect. This is the one property the counter
 * exists for, and it is checked through janet_collect rather than by reading
 * the field back. */
static void test_gclock_suspends_collection(void) {
    int handle = janet_gclock();
    janet_vm.next_collection = 12345;
    janet_collect();
    assert(janet_vm.next_collection == 12345);
    janet_gcunlock(handle);
    janet_collect();
    assert(janet_vm.next_collection == 0);
}

/* --------------------------------------------------------------- scratch */

static int finalizer_calls;
static void *finalizer_args[8];

static void record_finalizer(void *mem) {
    if (finalizer_calls < 8) finalizer_args[finalizer_calls] = mem;
    finalizer_calls++;
}

/* A scratch block is registered in the table, and the pointer handed back sits
 * exactly one header above it. That relationship is the whole allocator:
 * janet_srealloc, janet_sfree and janet_sfinalizer all recover the header by
 * subtracting from the pointer the caller holds. */
static void test_smalloc_registers(void) {
    size_t base = janet_vm.scratch_len;

    char *p = janet_smalloc(40);
    assert(p != NULL);
    assert(janet_vm.scratch_len == base + 1);
    assert(janet_vm.scratch_mem[base] == header_of(p));
    assert(header_of(p)->finalize == NULL);
    assert(((uintptr_t) p) % sizeof(long long) == 0);

    memset(p, 'x', 40);
    janet_sfree(p);
    assert(janet_vm.scratch_len == base);
}

/* janet_scalloc zeroes, and rejects a product that would wrap. The zero-length
 * cases still produce a registered block. */
static void test_scalloc(void) {
    size_t base = janet_vm.scratch_len;

    unsigned char *p = janet_scalloc(9, 7);
    assert(janet_vm.scratch_len == base + 1);
    for (size_t i = 0; i < 63; i++) assert(p[i] == 0);

    void *empty = janet_scalloc(0, 16);
    assert(empty != NULL);
    assert(janet_vm.scratch_len == base + 2);

    void *empty2 = janet_scalloc(16, 0);
    assert(empty2 != NULL);
    assert(janet_vm.scratch_len == base + 3);

    janet_sfree(empty2);
    janet_sfree(empty);
    janet_sfree(p);
    assert(janet_vm.scratch_len == base);
}

/* janet_srealloc keeps the block in the same table slot, preserves the bytes
 * that fit, and carries the finalizer across -- the header moves with the
 * allocation. A null pointer means allocate. */
static void test_srealloc(void) {
    size_t base = janet_vm.scratch_len;

    void *fresh = janet_srealloc(NULL, 24);
    assert(fresh != NULL);
    assert(janet_vm.scratch_len == base + 1);
    assert(scratch_index_of(fresh) == (int) base);
    janet_sfree(fresh);

    char *p = janet_smalloc(16);
    memcpy(p, "0123456789abcde", 16);
    janet_sfinalizer(p, record_finalizer);
    int slot = scratch_index_of(p);
    assert(slot >= 0);

    char *grown = janet_srealloc(p, 4096);
    assert(janet_vm.scratch_len == base + 1);
    assert(scratch_index_of(grown) == slot);
    assert(0 == memcmp(grown, "0123456789abcde", 16));
    assert(header_of(grown)->finalize == record_finalizer);

    char *shrunk = janet_srealloc(grown, 8);
    assert(janet_vm.scratch_len == base + 1);
    assert(scratch_index_of(shrunk) == slot);
    assert(0 == memcmp(shrunk, "01234567", 8));

    finalizer_calls = 0;
    janet_sfree(shrunk);
    assert(finalizer_calls == 1);
    assert(finalizer_args[0] == shrunk);
    assert(janet_vm.scratch_len == base);
}

/* Freeing fills the vacated table slot from the top, the same way the root set
 * does, and a null pointer is a no-op. */
static void test_sfree(void) {
    size_t base = janet_vm.scratch_len;

    void *a = janet_smalloc(8);
    void *b = janet_smalloc(8);
    void *cc = janet_smalloc(8);
    assert(janet_vm.scratch_len == base + 3);

    janet_sfree(NULL);
    assert(janet_vm.scratch_len == base + 3);

    janet_sfree(a);
    assert(janet_vm.scratch_len == base + 2);
    assert(scratch_index_of(cc) == (int) base);
    assert(scratch_index_of(b) == (int) base + 1);

    janet_sfree(b);
    janet_sfree(cc);
    assert(janet_vm.scratch_len == base);
}

/* A finalizer runs once, with the caller's pointer rather than the header. */
static void test_sfinalizer(void) {
    size_t base = janet_vm.scratch_len;

    void *p = janet_smalloc(8);
    finalizer_calls = 0;
    janet_sfree(p);
    assert(finalizer_calls == 0);

    p = janet_smalloc(8);
    janet_sfinalizer(p, record_finalizer);
    finalizer_calls = 0;
    janet_sfree(p);
    assert(finalizer_calls == 1);
    assert(finalizer_args[0] == p);
    assert(janet_vm.scratch_len == base);
}

/* The table grows to twice what is needed plus two, and everything already in
 * it survives the move. */
static void test_scratch_capacity_growth(void) {
    void *held[64];
    size_t base = janet_vm.scratch_len;
    size_t held_count = 0;

    while (janet_vm.scratch_len < janet_vm.scratch_cap) {
        held[held_count] = janet_smalloc(8);
        memset(held[held_count], (int) held_count, 8);
        held_count++;
        assert(held_count < 64);
    }
    size_t at_capacity = janet_vm.scratch_cap;

    held[held_count] = janet_smalloc(8);
    memset(held[held_count], (int) held_count, 8);
    held_count++;
    assert(janet_vm.scratch_cap == 2 * at_capacity + 2);
    assert(janet_vm.scratch_len == base + held_count);

    for (size_t i = 0; i < held_count; i++) {
        unsigned char *bytes = held[i];
        assert(scratch_index_of(held[i]) >= 0);
        for (size_t j = 0; j < 8; j++) assert(bytes[j] == (unsigned char) i);
    }
    for (size_t i = 0; i < held_count; i++) janet_sfree(held[i]);
    assert(janet_vm.scratch_len == base);
}

/* Releasing everything runs each finalizer and empties the table. This is what
 * janet_collect does at the end of a collection and janet_clear_memory does at
 * shutdown, which is why the scratch API needs no explicit free to be correct. */
static void test_free_all_scratch(void) {
    janet_collect();
    assert(janet_vm.scratch_len == 0);

    void *a = janet_smalloc(8);
    void *b = janet_smalloc(8);
    void *cc = janet_smalloc(8);
    janet_sfinalizer(a, record_finalizer);
    janet_sfinalizer(cc, record_finalizer);

    finalizer_calls = 0;
    janet_free_all_scratch();
    assert(janet_vm.scratch_len == 0);
    assert(finalizer_calls == 2);
    assert(finalizer_args[0] == a);
    assert(finalizer_args[1] == cc);
    (void) b;
}

/* And a collection does the same, on its way out. */
static void test_collect_frees_scratch(void) {
    void *p = janet_smalloc(8);
    janet_sfinalizer(p, record_finalizer);
    finalizer_calls = 0;
    janet_collect();
    assert(finalizer_calls == 1);
    assert(finalizer_args[0] == p);
    assert(janet_vm.scratch_len == 0);
}

int main(void) {
    janet_init();

    test_gcpressure();
    test_gcalloc_normal_heap();
    test_gcalloc_weak_heap();
    test_gcalloc_strong_types();
    test_gcalloc_chains();

    test_root_counting();
    test_root_identity_by_pointer();
    test_root_identity_of_immediates();
    test_root_removal_swaps_from_top();
    test_root_unrootall_halves();
    test_root_unrootall_absent();
    test_root_capacity_growth();

    test_gclock_nesting();
    test_gclock_suspends_collection();

    test_smalloc_registers();
    test_scalloc();
    test_srealloc();
    test_sfree();
    test_sfinalizer();
    test_scratch_capacity_growth();
    test_free_all_scratch();
    test_collect_frees_scratch();

    janet_deinit();
    printf("gc alloc contract ok\n");
    return 0;
}

File: src/zig/subsystems/gc_alloc.zig
//! jump-transparent
//!
//! The collector's memory: allocating a collectable block onto one of the two
//! heap lists, the root set, the GC suspend counter, and the scratch
//! allocator. This is the first of the three increments `gc.c` is split into;
//! marking and sweeping stay in C until Parts 4 and 5.
//!
//! Nothing here traverses or frees a collectable object. `janet_gcalloc`
//! returns raw memory with only its type tag written, exactly as the C
//! original does — the caller initialises the block and the collector never
//! looks at it until it is reachable. The block lists are touched at one end
//! only: this file prepends, and `janet_sweep` in `gc.c` unlinks.
//!
//! **The file is jump-transparent**, under the rule SPIKE-8 settled. One call
//! here reaches code this runtime does not own: `freeOneScratch` calls a
//! `JanetScratchFinalizer` an embedder installed through `janet_sfinalizer`.
//! No in-tree caller installs one, and a finalizer that raises is undefined
//! behaviour, but the frames below must still hold nothing that a skipped
//! cleanup would strand — so there is no `defer` in this file and `build.zig`
//! checks that there is not. A signal from a scratch finalizer leaves
//! `scratch_len` unreduced and the block re-finalized on the next collection,
//! which is what the C original does too; the port does not diverge there.
//!
//! Two pieces of the C arithmetic are reproduced rather than repaired, and
//! both are in `FOUND.md`:
//!
//!  - `janet_smalloc` grows the scratch table by `newcap * sizeof(JanetScratch)`
//!    where the element is a `JanetScratch *`. The header is a function pointer
//!    plus a flexible array, so it is never smaller than a pointer and the
//!    table is over-allocated rather than short — harmless, and preserved so
//!    the two selectors request the same byte counts.
//!  - `janet_smalloc` and `janet_srealloc` add the header size to the caller's
//!    size without checking for wraparound, so a near-`SIZE_MAX` request
//!    allocates a few bytes and returns a pointer into a block far too small.
//!    Wrapping addition is used below to reproduce it exactly rather than trap.
//!
//! One assumption the C code makes is worth naming because this file relies on
//! it in the same way. `janet_smalloc` returns `s->mem` and `janet_mem2scratch`
//! recovers the header with `((JanetScratch *)mem) - 1`; those are inverses
//! only if `sizeof(JanetScratch)` equals `offsetof(JanetScratch, mem)`, which
//! holds because the header is one pointer followed by a `long long[]`. Zig
//! does the same arithmetic through `header_size` below, so the two
//! implementations hand out and accept the same addresses.

const std = @import("std");
const abi = @import("abi");
const c = abi.c;

/// `janet_vm` as `src/core/state.h` declares it, resolved through `abi.zig`.
inline fn vm() *c.JanetVM {
    return &c.janet_vm;
}

/// The first memory type that belongs on the weak heap. `gc.c` compares
/// against the enum constant directly; naming it here keeps the comparison in
/// the enum's own type, so a caller passing a value outside the enumeration
/// lands where C lands rather than tripping a conversion check.
const first_weak_type: c.enum_JanetMemoryType = c.JANET_MEMORY_TABLE_WEAKK;

/// The distance from a `JanetScratch` header to the memory it hands out. See
/// the note at the head of the file about why this is `@sizeOf` and not an
/// offset of the flexible array member.
const header_size = @sizeOf(c.JanetScratch);

// ------------------------------------------------------------- collectable

/// Tell the collector that `s` bytes were allocated outside its accounting.
/// Unsigned wraparound is defined in C and reproduced here; the field is a
/// heuristic threshold, and a wrapped value only delays a collection.
export fn janet_gcpressure(s: usize) callconv(.c) void {
    vm().next_collection +%= s;
}

/// Allocate a block the collector owns and prepend it to the appropriate heap
/// list. The block is *not* initialised beyond its type tag, and it is
/// reachable by the collector from this moment, so every caller fills it in
/// before anything can collect.
export fn janet_gcalloc(mtype: c.enum_JanetMemoryType, size: usize) callconv(.c) ?*anyopaque {
    const v = vm();

    if (v.cache == null) c.janet_zig_fatal("please initialize janet before use");

    const mem: *c.JanetGCObject = @ptrCast(@alignCast(c.janet_malloc(size) orelse
        c.janet_zig_out_of_memory()));

    mem.flags = @bitCast(@as(u32, @truncate(mtype)));

    v.next_collection +%= size;
    if (mtype < first_weak_type) {
        mem.data.next = @ptrCast(@alignCast(v.blocks));
        v.blocks = mem;
    } else {
        mem.data.next = @ptrCast(@alignCast(v.weak_blocks));
        v.weak_blocks = mem;
    }
    v.block_count +%= 1;

    return mem;
}

// -------------------------------------------------------------- root set

/// Add a root. Rooting the same value twice requires unrooting it twice; the
/// root set is a multiset, not a set, which is why this appends unconditionally.
export fn janet_gcroot(root: c.Janet) callconv(.c) void {
    const v = vm();
    const newcount = v.root_count +% 1;
    if (newcount > v.root_capacity) {
        const newcap = 2 *% newcount;
        // The C original stores the result before testing it, so a failed
        // grow leaves `roots` null on the way to exiting. Preserved.
        v.roots = @ptrCast(@alignCast(c.janet_realloc(v.roots, @sizeOf(c.Janet) *% newcap)));
        if (v.roots == null) c.janet_zig_out_of_memory();
        v.root_capacity = newcap;
    }
    v.roots[v.root_count] = root;
    v.root_count = newcount;
}

/// Identity for root bookkeeping. The three immediate types compare equal to
/// any other value of their type, which costs nothing: the collector never
/// traces them, so which one a root slot holds cannot matter.
fn idequals(lhs: c.Janet, rhs: c.Janet) bool {
    if (c.janet_type(lhs) != c.janet_type(rhs)) return false;
    return switch (c.janet_type(lhs)) {
        c.JANET_BOOLEAN, c.JANET_NIL, c.JANET_NUMBER => true,
        else => c.janet_unwrap_pointer(lhs) == c.janet_unwrap_pointer(rhs),
    };
}

/// Drop one rooting of `root`, returning whether one was found. Removal swaps
/// the last root into the vacated slot, so the root set has no order a caller
/// may depend on.
export fn janet_gcunroot(root: c.Janet) callconv(.c) c_int {
    const v = vm();
    const top = v.root_count;
    // Bottom to top, as the C original is; its comment says the access
    // pattern is expected to be LIFO, but the scan starts at slot zero.
    var i: usize = 0;
    while (i < top) : (i += 1) {
        if (idequals(root, v.roots[i])) {
            v.root_count -%= 1;
            v.roots[i] = v.roots[v.root_count];
            return 1;
        }
    }
    return 0;
}

/// Drop every rooting of `root`, returning whether there was at least one.
///
/// It does not, in fact, drop every one. Filling the vacated slot from the top
/// and then advancing skips whatever was moved down, so a root that appears
/// twice can survive with one rooting left. `FOUND.md` records it; this
/// reproduces it. `top` shadows `root_count` the way the C original's `vtop`
/// shadows `roots + root_count` — the two fall together, one per match.
export fn janet_gcunrootall(root: c.Janet) callconv(.c) c_int {
    const v = vm();
    var top = v.root_count;
    var ret: c_int = 0;
    var i: usize = 0;
    while (i < top) : (i += 1) {
        if (idequals(root, v.roots[i])) {
            v.root_count -%= 1;
            v.roots[i] = v.roots[v.root_count];
            top -= 1;
            ret = 1;
        }
    }
    return ret;
}

// ------------------------------------------------------------ suspension

/// Suspend collection, returning the previous nesting depth. The handle is the
/// depth to restore, not a token to match, so unlocking with a stale handle
/// deliberately unwinds every lock taken since it was issued.
export fn janet_gclock() callconv(.c) c_int {
    const v = vm();
    const previous = v.gc_suspend;
    v.gc_suspend = previous +% 1;
    return previous;
}

export fn janet_gcunlock(handle: c_int) callconv(.c) void {
    vm().gc_suspend = handle;
}

// --------------------------------------------------------------- scratch

/// The memory a scratch header hands out.
inline fn scratchData(s: *c.JanetScratch) *anyopaque {
    return @ptrFromInt(@intFromPtr(s) + header_size);
}

/// The inverse. Wrapping subtraction so that a null argument produces the same
/// wild pointer the C original computes rather than trapping before it.
inline fn mem2scratch(mem: ?*anyopaque) *c.JanetScratch {
    return @ptrFromInt(@intFromPtr(mem) -% header_size);
}

/// Run a scratch block's finalizer, if it has one, and release it. The
/// finalizer is embedder code; see the jump-transparency note at the head of
/// the file.
fn freeOneScratch(s: *c.JanetScratch) void {
    if (s.finalize) |finalize| finalize(scratchData(s));
    c.janet_free(s);
}

/// Release every scratch block. Called by `janet_collect` at the end of a
/// collection and by `janet_clear_memory` at shutdown, both still in `gc.c`;
/// declared in `gc.h` so that this file can provide it for them.
export fn janet_free_all_scratch() callconv(.c) void {
    const v = vm();
    var i: usize = 0;
    while (i < v.scratch_len) : (i += 1) freeOneScratch(v.scratch_mem[i]);
    v.scratch_len = 0;
}

/// Allocate scratch memory: freed automatically at the next collection, and
/// optionally before that with `janet_sfree`. The header carries the finalizer
/// and the table of live blocks carries the pointer.
export fn janet_smalloc(size: usize) callconv(.c) ?*anyopaque {
    const s: *c.JanetScratch = @ptrCast(@alignCast(c.janet_malloc(header_size +% size) orelse
        c.janet_zig_out_of_memory()));
    s.finalize = null;

    const v = vm();
    if (v.scratch_len == v.scratch_cap) {
        const newcap = 2 *% v.scratch_cap +% 2;
        // `header_size` rather than the pointer size, reproducing the C
        // original's element size. See the note at the head of the file.
        // The cast is only Zig's: a `JanetScratch **` is a double pointer,
        // which does not coerce to `void *` on its own.
        const newmem = c.janet_realloc(@ptrCast(v.scratch_mem), newcap *% header_size) orelse
            c.janet_zig_out_of_memory();
        v.scratch_cap = newcap;
        v.scratch_mem = @ptrCast(@alignCast(newmem));
    }

    v.scratch_mem[v.scratch_len] = s;
    v.scratch_len +%= 1;
    return scratchData(s);
}

export fn janet_scalloc(nmemb: usize, size: usize) callconv(.c) ?*anyopaque {
    if (nmemb != 0 and size > std.math.maxInt(usize) / nmemb) c.janet_zig_out_of_memory();
    const n = nmemb *% size;
    const p = janet_smalloc(n).?;
    @memset(@as([*]u8, @ptrCast(p))[0..n], 0);
    return p;
}

/// Resize a scratch block in place in the table. A pointer that is not a live
/// scratch block is fatal rather than diagnosable, as in the C original: the
/// table is the only record that a block exists, and a miss means the caller
/// passed something this allocator never handed out.
export fn janet_srealloc(mem: ?*anyopaque, size: usize) callconv(.c) ?*anyopaque {
    if (mem == null) return janet_smalloc(size);
    const s = mem2scratch(mem);
    const v = vm();
    var i = v.scratch_len;
    while (i > 0) {
        i -= 1;
        if (v.scratch_mem[i] == s) {
            const news: *c.JanetScratch = @ptrCast(@alignCast(c.janet_realloc(s, size +% header_size) orelse
                c.janet_zig_out_of_memory()));
            v.scratch_mem[i] = news;
            return scratchData(news);
        }
    }
    c.janet_zig_fatal("invalid janet_srealloc");
}

/// Install a finalizer to run when the block is released, whether by
/// `janet_sfree` or by the next collection.
export fn janet_sfinalizer(mem: ?*anyopaque, finalizer: c.JanetScratchFinalizer) callconv(.c) void {
    mem2scratch(mem).finalize = finalizer;
}

export fn janet_sfree(mem: ?*anyopaque) callconv(.c) void {
    if (mem == null) return;
    const s = mem2scratch(mem);
    const v = vm();
    var i = v.scratch_len;
    while (i > 0) {
        i -= 1;
        if (v.scratch_mem[i] == s) {
            v.scratch_len -%= 1;
            v.scratch_mem[i] = v.scratch_mem[v.scratch_len];
            freeOneScratch(s);
            return;
        }
    }
    c.janet_zig_fatal("invalid janet_sfree");
}
